pennylane/devices/qubit/preprocess.py

```diff
@@ -95,34 +100,18 @@ def validate_and_expand_adjoint(
         Union[.QuantumTape, .DeviceError]: The expanded tape, such that it is supported by adjoint differentiation.
         If the circuit is invalid for adjoint differentiation, a DeviceError with an explanation is returned instead.
     """
-    # Check validity of measurements
-    measurements = []
-    for m in circuit.measurements:
-        if not isinstance(m, ExpectationMP):
-            return DeviceError(
-                "Adjoint differentiation method does not support "
-                f"measurement {m.__class__.__name__}."
-            )
-
-        if not m.obs.has_matrix:
-            return DeviceError(
-                f"Adjoint differentiation method does not support observable {m.obs.name}."
-            )
-
-        measurements.append(m)
 
-    expanded_ops = []
-    for op in circuit._ops:
-        if op.num_params > 1:
-            if not isinstance(op, qml.Rot):
-                return DeviceError(
-                    f"The {op} operation is not supported using "
-                    'the "adjoint" differentiation method.'
-                )
-            ops = op.decomposition()
-            expanded_ops.extend(ops)
-        elif not isinstance(op, qml.operation.StatePrep):
-            expanded_ops.append(op)
+    try:
+        new_ops = [
+            final_op
+            for op in circuit._ops
+            for final_op in _operator_decomposition_gen(op, _accepted_adjoint_operator)
+        ]
```

Comment:
The reason why previously we only allowed `Rot` to be decomposed and not any other multi-parameter operator is because the decomposition of `Rot` has the same parameters as the ones that are passed in. This is not true for the decompositions of other parametrized operators.

For example, `U2` takes two parameters, so it would get decomposed. However, the operators from the decomposition collectively accept 5 parameters, which would then become the parameters that are differentiated.

It could potentially be fixed with the addition of some type of quantum to classical jacobian post-processing, which is honestly not really worth the trouble in my opinion. I don't even know how we would track the derivatives between the actual parameters and the decomposed parameters in a general way.

Reply:
So the intention of `Device.preprocess` is that its tracked by the machine learning interface. Therefore it can have a classical component to its transformations, see [`execution.py:_adjoint_jacobian_expansion`](https://github.com/PennyLaneAI/pennylane/blob/8af85140f9778e9a171e0e1bf8ebf6cbab42222d/pennylane/interfaces/execution.py#L66). We are basically wanting to get rid of the patch by moving device specific expansion to Device preprocessing.

If we were performing this expansion inside `compute_derivatives` somewhere, we could not decompose `U3`. But since this function is tracked by the machine learning interface, we can have classical components to the derivatives.

Reply:
> It could potentially be fixed with the addition of some type of quantum to classical jacobian post-processing, which is honestly not really worth the trouble in my opinion. I don't even know how we would track the derivatives between the actual parameters and the decomposed parameters in a general way.

@mudit2812 I'm curious to understand more here 🤔 As long as the processing of the parameters during decomposition is done using AD functions (e.g., `qml.math`), _and_ the decomposition happens before the `custom_vjp`/interface boundary, everything should 'just work'.

Reply:
Thanks for the context! There's a lot about derivative computation in Pennylane that I'm very unfamiliar with. I have more clarity on it now :)